builtin/providers/aws/resource_aws_ssm_document.go

```diff
@@ -0,0 +1,309 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsSsmDocument() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSsmDocumentCreate,
+		Read:   resourceAwsSsmDocumentRead,
+		Update: resourceAwsSsmDocumentUpdate,
+		Delete: resourceAwsSsmDocumentDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"content": &schema.Schema{
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"created_date": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"hash": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"hash_type": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"owner": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"status": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"platform_type": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"parameter": &schema.Schema{
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"default_value": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"description": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"type": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"permissions": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"type": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"account_ids": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsSsmDocumentCreate(d *schema.ResourceData, meta interface{}) error {
+	ssmconn := meta.(*AWSClient).ssmconn
+
+	log.Printf("[INFO] Creating SSM Document: %s", d.Get("name").(string))
+
+	docInput := &ssm.CreateDocumentInput{
+		Name:    aws.String(d.Get("name").(string)),
+		Content: aws.String(d.Get("content").(string)),
+	}
+
+	resp, err := ssmconn.CreateDocument(docInput)
+
+	if err != nil {
+		return fmt.Errorf("[ERROR] Error creating SSM document: %s", err)
```

Comment:
Wherever we return an error which originates from upstream, the current best practice is to wrap it using the `errwrap` library. So in this case, this line would read:

```go
return errwrap.Wrapf("[ERROR] Error creating SSM document: {{err}}", err)
```

This preserves any context attached to the original error, as well as allowing tests for particular types of error having occurred in a given chain.